readthedocs/builds/models.py

```diff
@@ -690,3 +690,128 @@ def run_time(self):
         if self.start_time is not None and self.end_time is not None:
             diff = self.end_time - self.start_time
             return diff.seconds
+
+
+# class VersionAutomationRuleQuerySet(SubclassMixin, models.QuerySet):
+#
+#     """Returns a subclass of VersionAutomationRule, based on the rule type."""
+#
+#     subclass_field = 'rule_type'
+#     subclass_field_id = 'rule_type_id'
+
+
+class VersionAutomationRule(TimeStampedModel, models.Model):
+
+    """Versions automation rules for projects."""
+
+    REGEX_RULE = 'regex-rule'
+    RULE_TYPES = (
+        (REGEX_RULE, _('Regular expression rule')),
+    )
+
+    ACTIVATE_VERSION_ACTION = 'activate-version'
+    ACTIONS = (
+        (ACTIVATE_VERSION_ACTION, _('Activate version on match')),
+    )
+
+    project = models.ForeignKey(
+        Project,
+        related_name='automation_rules',
+        on_delete=models.CASCADE,
+    )
+    priority = models.IntegerField(_('Rule priority'))
+    rule_type = models.CharField(
+        _('Rule type'),
+        max_length=32,
+        choices=RULE_TYPES,
+    )
+    rule_arg = models.CharField(
+        _('Value used for the rule to match the version'),
+        max_length=255,
+    )
+    action = models.CharField(
+        _('Action'),
+        max_length=32,
+        choices=ACTIONS,
+    )
+    action_arg = models.CharField(
+        _('Value used for the action to perfom an operation'),
+        max_length=255,
+        null=True,
+        blank=True,
+    )
+    version_type = models.CharField(
+        _('Version type'),
+        max_length=32,
+        choices=VERSION_TYPES,
+    )
+
+    class Meta:
+        unique_together = (('project', 'priority'),)
+        ordering = ('priority', '-modified', '-created')
+
+    def run(self, version, *args, **kwargs):
+        """
+        Run an action if `version` matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        """
+        if version.type == self.version_type:
+            match_result = self.match(version, self.rule_arg)
+            if match_result is not None:
+                return self.apply_action(version, match_result, self.action_arg)
+        return False
+
+    def match(self, version, arg):
```

Comment:
I'm thinking is passing `*args` and `**kwargs` to match and apply_action, still not sure. And maybe select a better name for `arg`. Or just use `self.rule_arg`, but this makes it easier to test :)

Reply:
I think it could be better to use `self.rule_arg` inside the method because it will more clear when reading the code and it's a common pattern.